### PR TITLE
feat: Maxcompute Sink Converters Refactor + Performance Improvement

### PR DESCRIPTION
**Pull request by ekawinataa**
Changes: 

- Consolidate TypeInfoConverter and PayloadConverter to a single interface
- Introduce TypeInfo caching and map based converter lookup  in form of MaxComputeProtobufConverterCache
- Fix performance issue due frequent access to aeonbit owner configuration object in conversion process
- Add instrumentation on streaming insert manager
- Remove MaxComputeClient dependency from MaxComputeSink. Previously, same StreamUploadSessions were used for all thread. Now, every MaxComputeSink are given individual InsertManager on initialization
- Add flag to enable default value on null fields for MaxCompute Proto Payload
- Bump odps lib version to 0.51.0-public

Before Improvement: 
- Conversion process took almost 92% of the overall processing time, leaving the rest for the insertion/flushing operation. After we profiled the application, it was found that using `config.someProps` directly introduced significant performance overhead due to heavy reflection used by Aeonbit Owner library

image

After Improvement: 
- Assigning configurations as individual class property and referring it from the conversin process improves the performance significantly

image

### CODE REVIEW

File: src/main/java/com/gotocompany/depot/metrics/MaxComputeMetrics.java [hunk 1]
```diff
@@ -37,4 +37,7 @@ public String getMaxComputeFlushRecordMetric() {
         return String.format("%s%s%s%s", getApplicationPrefix(), SINK_PREFIX, MAXCOMPUTE_SINK_PREFIX, "flush_record_count");
     }
 
+    public String getMaxComputeConversionLatencyMetric() {
```
Comment (mayankrai09) [inline src/main/java/com/gotocompany/depot/metrics/MaxComputeMetrics.java:40]: Can we also add metrics to track the number of upload sessions created by firehose?

File: src/main/java/com/gotocompany/depot/maxcompute/converter/MessageProtobufMaxComputeConverter.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    public boolean canConvert(Descriptors.FieldDescriptor fieldDescriptor) {
+        return Descriptors.FieldDescriptor.Type.MESSAGE.equals(fieldDescriptor.getType());
+    }
+
+    @Override
+    public Object convertSingularPayload(ProtoPayload protoPayload) {
+        Message dynamicMessage = (Message) protoPayload.getParsedObject();
+        List<Object> values = new ArrayList<>();
+        Map<Descriptors.FieldDescriptor, Object> payloadFields = dynamicMessage.getAllFields();
+        protoPayload.getFieldDescriptor().getMessageType().getFields().forEach(innerFieldDescriptor -> {
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/depot/maxcompute/converter/MessageProtobufMaxComputeConverter.java:65]: revert

File: src/main/java/com/gotocompany/depot/maxcompute/converter/ProtobufMaxComputeConverter.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    TypeInfo convertSingularTypeInfo(Descriptors.FieldDescriptor fieldDescriptor);
+
+    /**
+     * Wraps the singular TypeInfo with array type handling if the field is repeated.
+     *
+     * @param fieldDescriptor the Protobuf field descriptor
+     * @param typeInfo the singular TypeInfo to wrap
+     * @return the wrapped TypeInfo, handling repeated fields as arrays
+     */
+    default TypeInfo wrapTypeInfo(Descriptors.FieldDescriptor fieldDescriptor, TypeInfo typeInfo) {
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/depot/maxcompute/converter/ProtobufMaxComputeConverter.java:40]: simplify

File: src/main/java/com/gotocompany/depot/maxcompute/MaxComputeSinkFactory.java [hunk 4]
```diff
@@ -67,10 +67,12 @@ public Sink create() {
                 messageParser,
                 partitioningStrategy,
                 maxComputeSinkConfig,
-                sinkConfig
+                sinkConfig,
+                new Instrumentation(statsDReporter, RecordDecorator.class),
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/depot/maxcompute/MaxComputeSinkFactory.java:71]: pass the StatsDReporter and let the concerned class instantiate the instrumentation

File: src/main/java/com/gotocompany/depot/maxcompute/MaxComputeSinkFactory.java [hunk 5]
```diff
@@ -67,10 +67,12 @@ public Sink create() {
...
+                maxComputeMetrics
         );
         ProtoMessageRecordConverter protoMessageRecordConverter = new ProtoMessageRecordConverter(recordDecorator, maxComputeSchemaCache);
-        return new MaxComputeSink(maxComputeClient, protoMessageRecordConverter,
+        return new MaxComputeSink(maxComputeClient.createInsertManager(), protoMessageRecordConverter,
                 new Instrumentation(statsDReporter, MaxComputeSink.class), maxComputeMetrics);
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/depot/maxcompute/MaxComputeSinkFactory.java:76]: pass the StatsDReporter and let the concerned class instantiate the instrumentation

File: src/main/java/com/gotocompany/depot/maxcompute/client/insert/InsertManagerFactory.java [hunk 6]
```diff
@@ -6,19 +6,42 @@
...
+     * @param maxComputeMetrics    metrics for MaxCompute
+     * @return InsertManager
+     */
     public static InsertManager createInsertManager(MaxComputeSinkConfig maxComputeSinkConfig,
                                                     TableTunnel tableTunnel,
                                                     Instrumentation instrumentation,
                                                     MaxComputeMetrics maxComputeMetrics) {
-        StreamingSessionManager streamingSessionManager = maxComputeSinkConfig.isTablePartitioningEnabled()
-                ? StreamingSessionManager.createPartitioned(tableTunnel, maxComputeSinkConfig) : StreamingSessionManager.createNonPartitioned(tableTunnel, maxComputeSinkConfig);
+        StreamingSessionManager streamingSessionManager = getStreamingSessionManager(maxComputeSinkConfig, tableTunnel, instrumentation, maxComputeMetrics);
```
Comment (ekawinataa) [inline src/main/java/com/gotocompany/depot/maxcompute/client/insert/InsertManagerFactory.java:29]: revert to ternary

Comment (ekawinataa) [reply src/main/java/com/gotocompany/depot/maxcompute/client/insert/InsertManagerFactory.java:29]: Declare it directly under the insert manager

**Review by rajuGT — CHANGES_REQUESTED** [review]

**Review by rajuGT — APPROVED** [review]

File: docs/reference/configuration/generic.md [hunk 7]
```diff
@@ -97,6 +97,13 @@ Scenario:
 * Type: `required`
 * Default: `MESSAGE`
 
+## `SINK_CONNECTOR_SCHEMA_PROTO_UNKNOWN_FIELDS_VALIDATION_INSTRUMENTATION_ENABLE`
+
+Enable instrumentation for unknown fields validation. Instrumentation itself will introduces some overhead, so it should be enabled when needed.
```
Comment (mayankrai09) [inline docs/reference/configuration/generic.md:102]: Enable instrumentation for unknown field validation for individual records. The instrumentation will introduce some overhead, So it should be enabled when needed.

**Review by mayankrai09 — APPROVED** [review]

**Pull request merged**